include/crab/domains/symbolic_variable_eq_domain.hpp

```diff
@@ -0,0 +1,1031 @@
+  void assign_bool_var(const element_t &lhs, const element_t &rhs,
+                       bool is_not_rhs) {
+    // WARN: this should not happen
+    return;
+  }
+
+  /// @brief remove an element from a class
+  /// @param v an element
+  /// @note  if v \in cls \land v is representative, select a new representative
+  ///        before removing v.
+  void operator-=(const element_t &v) {
+    SVEQ_DOMAIN_SCOPED_STATS(".forget");
+    if (is_bottom() || is_top()) {
+      return;
+    }
+    CRAB_LOG("symb-var-eq", crab::outs() << "Forgetting " << v << ": ";
+             dump(););
+
+    if (contains(v)) {
+      // v is not a representative, jusrt remove v
+      // if not, pick a new one and update all members
+      if (m_parents.at(v) == v) {
+        // v is the representative of the equivalence class
+        boost::optional<element_t> new_rep;
+        for (auto &kv : m_parents) {
+          // search element which is not v but in v's class
+          if (kv.first != v && kv.second == v) {
+            if (!new_rep) {
+              // choose the first one as representative
+              new_rep = kv.first;
+              m_classes.insert({*new_rep, m_classes.at(v)});
+            }
+            m_parents.at(kv.first) = *new_rep;
+          }
+        }
+        m_classes.erase(v);
+      }
+      m_parents.erase(v);
+    }
+    normalize();
+    CRAB_LOG("symb-var-eq", crab::outs() << "After forget "; dump(););
+  }
+
+  /// @brief alternative operation to forget a set of elements
+  /// @param elements a set of elements
+  void forget(const std::vector<element_t> &elements) {
+    if (is_bottom() || is_top()) {
+      return;
+    }
+
+    for (auto v : elements) {
+      *this -= v;
+    }
+  }
+
+  /// @brief keep equivalence classes only for element in vector elements
+  /// @param elements a vector of elements required to keep
+  /// @note  After project operation, the state only keeps equivalence classes
+  ///        for element in vector elements
+  void project(const std::vector<element_t> &elements) {
+    SVEQ_DOMAIN_SCOPED_STATS(".project");
+
+    if (is_bottom() || is_top()) {
+      return;
+    }
+    CRAB_LOG("symb-var-eq", crab::outs() << "Projecting ";
+             print_elems_vector(crab::outs(), elements); crab::outs() << "\n";
+             dump(););
+    // group elements based on current equivalence classes
+    equivalence_class_elems_t elems;
+
+    this_domain_t res;
+
+    for (auto &v : elements) {
+      if (contains(v)) {
+        element_t rep = find(v);
+        element_set_t &s = elems[rep];
+        s.push_back(v);
+      }
+    }
+
+    for (auto &kv : elems) {
+      const element_t &rep = kv.first;
+      const element_set_t &s = kv.second;
+      bool keep_rep = false;
+      for (auto &e : s) {
+        if (e == rep) {
+          keep_rep = true;
+          break;
+        }
+      }
+      element_t new_rep = keep_rep ? rep : s[0];
```

Comment:
Why create new representatives? Can't you just forget what is not in `elements`?